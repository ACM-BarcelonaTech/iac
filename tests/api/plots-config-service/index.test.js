const AWSMock = require('aws-sdk-mock');
const AWS = require('aws-sdk');
const handlePlotConfigRequest = require('../../../src/api/plots-config-service');

jest.mock('../../../src/config');
jest.mock('../../../src/utils/logging');

const socket = jest.fn();
const emit = jest.fn();
socket.emit = emit;


describe('Test Plot Config Service', () => {
  afterEach(() => {
    socket.mockClear();
    emit.mockClear();
    AWSMock.restore('DynamoDB');
  });

  it('create operation, puts data in database', async () => {
    const workRequest = {
      uuid: '1',
      operation: 'create',
      plotName: 'plot1',
      experimentId: '1',
      body: { foo: 'bar', baz: [1, 2, 3] },
    };

    const putItem = jest.fn();
    AWSMock.setSDKInstance(AWS);
    AWSMock.mock('DynamoDB', 'putItem', (params, callback) => {
      putItem(params);
      callback(null, null);
    });
    await handlePlotConfigRequest(workRequest, socket);
    expect(putItem).toMatchSnapshot();
  });
  it('delete operation, delete data from database', async () => {
    const workRequest = {
      uuid: '1',
      operation: 'delete',
      plotName: 'plot1',
      experimentId: '1',
    };

    const deleteItem = jest.fn();
    AWSMock.setSDKInstance(AWS);
    AWSMock.mock('DynamoDB', 'deleteItem', (params, callback) => {
      deleteItem(params);
      callback(null, null);
    });
    await handlePlotConfigRequest(workRequest, socket);
    expect(deleteItem).toMatchSnapshot();
  });
  it('update operation, update data in database', async () => {
    const workRequest = {
      uuid: '1',
      operation: 'update',
      plotName: 'plot1',
      experimentId: '1',
      body: { foo: 'bar', baz: [1, 2, 3] },
    };

    const updateItem = jest.fn();
    AWSMock.setSDKInstance(AWS);
    AWSMock.mock('DynamoDB', 'updateItem', (params, callback) => {
      updateItem(params);
      callback(null, null);
    });
    await handlePlotConfigRequest(workRequest, socket);
    expect(updateItem).toMatchSnapshot();
  });
  it('read operation, read data from database', async () => {
    const workRequest = {
      uuid: '1',
      operation: 'read',
      plotName: 'plot1',
      experimentId: '1',
    };

    const getItem = jest.fn();
    AWSMock.setSDKInstance(AWS);
    AWSMock.mock('DynamoDB', 'getItem', (params, callback) => {
      getItem(params);
      callback(null, null);
    });
    await handlePlotConfigRequest(workRequest, socket);
    expect(getItem).toMatchSnapshot();
  });
});
