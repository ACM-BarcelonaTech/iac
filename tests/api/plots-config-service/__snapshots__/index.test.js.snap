// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Plot Config Service create operation, puts data in database 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Item": Object {
          "data": Object {
            "M": Object {
              "baz": Object {
                "L": Array [
                  Object {
                    "N": "1",
                  },
                  Object {
                    "N": "2",
                  },
                  Object {
                    "N": "3",
                  },
                ],
              },
              "foo": Object {
                "S": "bar",
              },
            },
          },
          "id": Object {
            "S": "1-plot1",
          },
        },
        "TableName": "plots-configs",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Test Plot Config Service delete operation, delete data from database 1`] = `[MockFunction]`;

exports[`Test Plot Config Service read operation, read data from database 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Key": Object {
          "id": Object {
            "S": "1-plot1",
          },
        },
        "TableName": "plots-configs",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Test Plot Config Service update operation, update data in database 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ExpressionAttributeNames": Object {
          "#d": "data",
        },
        "ExpressionAttributeValues": Object {
          ":x": Object {
            "M": Object {
              "baz": Object {
                "L": Array [
                  Object {
                    "N": "1",
                  },
                  Object {
                    "N": "2",
                  },
                  Object {
                    "N": "3",
                  },
                ],
              },
              "foo": Object {
                "S": "bar",
              },
            },
          },
        },
        "Key": Object {
          "id": Object {
            "S": "1-plot1",
          },
        },
        "TableName": "plots-configs",
        "UpdateExpression": "set #d = :x",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
